### PR TITLE
infer non-empty-string on literal-string and strlen()

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

**Comment by staabm** [comment]
memory related errors are also present on latest 1.6.x branch, therefore feel unrelated

**Comment by staabm** [comment]
memory related errors are also present on latest 1.6.x branch, therefore feel unrelated

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -235,9 +237,19 @@ public function specifyTypesInCondition(
 							$newContext = $newContext->negate();
 						}
 						$argType = $scope->getType($exprNode->getArgs()[0]->value);
-						if ($argType instanceof StringType) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:238]: primary motivation was killing `instanceof StringType` checks

**Comment by staabm** [comment]
I am on holiday for a few days. Will investigate the webmozart/assert error after coming back...

**Comment by herndlm** [comment]
You actually fixed something see comment in test file there ;) I overdid it with some of the tests..

**Comment by staabm** [comment]
Even better :-)

File: src/Analyser/TypeSpecifier.php [hunk 2]
```diff
@@ -235,9 +237,19 @@ public function specifyTypesInCondition(
...
+						if ($argType->isString()->yes()) {
+							$accessories = [
+								new AccessoryNonEmptyStringType(),
+							];
```
Comment (herndlm) [inline src/Analyser/TypeSpecifier.php:243]: I guess it should not be a problem atm, but wouldn't it make more sense / be safer or more consistent to do something like the following instead?
```suggestion
                        $accessories = [];
						if ($argType->isString()->yes()) {
							$accessories[] = new AccessoryNonEmptyStringType();
```

Comment (herndlm) [reply src/Analyser/TypeSpecifier.php:243]: ah sorry, I misread the if indentation :)

**Comment by ondrejmirtes** [comment]
Just added a commit here: https://github.com/phpstan/phpstan-src/pull/1260/commits/2821732311661013c08b66fda3bcfbc209a3cd0c

The takeaway is - if something already is something, you don't need to tell TypeSpecifier to narrow it down again. So if you have strlen($literalString), TypeSpecifier doesn't need to say "add literal-string again".

**Comment by ondrejmirtes** [comment]
So the useful thing is:

```diff
-						if ($argType instanceof StringType) {
+						if ($argType->isString()->yes()) {
```

Which is great and should be done in more relevant places.

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by staabm** [comment]
> Which is great and should be done in more relevant places.

Yep, I am on it - after holiday :-)

**Pull request merged**